### ISSUE TITLE
adding strides, dilations and batches to convolutions ( feature request )

### ISSUE THREAD

**Issue by danFromTelAviv** [scipy/scipy#9295]
Much of the scientific work that is being done today is in neural networks. The use of convolutions is very significant there. I would like to see the scipy convolutions better fitted to this task while keeping backward compatibility for other scientific tasks. 

There is already nd convolutions which are almost all that is needed. The next step is to add strides, dilations ( atrous convolutions ) and the ability to process batches of images/ nd matrices in one call. 

Specifically, I am asking for this in order to [ add numpy/scipy as a backend for keras](https://github.com/keras-team/keras/issues/11068) 

I am working on a  [cython implementation](https://github.com/danFromTelAviv/cynfrence) and found that it is competitively fast compared to tensorflow on CPU however this really seems like something that would be a better fit as a part of scipy with easier access for everyone.

**Comment by ilayn** [comment]
I don't know exactly what the context is but can you please provide what the difference is, how it is going to be implemented or the benchmarks about the results.

Also SciPy uses FFTPACK and other tools in C in certain scenarios, so for those branches you can't get rid of the overhead (in case it is relevant). See the source code for more details.

**Comment by danFromTelAviv** [comment]
The context : 
You have a set of images that you want to convolve with a set of filters. This is the basis of a convolutional neural network and is also useful in other cases - for example as a base metric for biclustering. We often are also interested in striding and dilating the convolution : 
[pretty decent explanation](https://www.quora.com/What-is-the-difference-between-dilated-convolution-and-convolution+stride)

My vision:
I would like to have 4 more inputs for [convolve](https://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.ndimage.filters.convolve.html)
1+2. input_batch_dim, weights_batch_dim - specifies one of the dimensions as a dimension which is not convolved but instead iterated over. In the example above this would mean run over each image and convolved it with each filter. 
3. strides - a tuple which specifies the strides on each dimension. 
4. dilations - a tuple which specifies the dilation on each dimension.

Another nice feature would be for input_batch_dim and weights_batch_dim to have no padding by default since they are not convolved. 

Implementation:
Obviously, in some cases, the overhead would be hard to avoid. None the less it would reduce the user's code since often there are cases when you want to "cross convolve" 2 sets of matrices. It would be awesome if there was at least one option that implements this in the c code directly since there would also be significant speed gains. 

As i said I wrote some reference cython code. My code currently convolves a batch of 3d ( width, height, color/channel) images with a batch of 3d filters. 

Benchmarking:
benchmarking for convolutions can be done like tests would. Find the speed in comparison to other libraries on a set of matrices and configurations.  I am not quite sure if you meant something else?

**Comment by rgommers** [comment]
This feature request is very specific to deep learning. `input_batch_dim`, `weights_batch_dim`, etc. don't really make sense in the API of a general convolution function. Therefore I'd say this is out of scope of SciPy.

> Specifically, I am asking for this in order to add numpy/scipy as a backend for keras

I've read through that issue. It makes sense to have a "basic" backend with good ARM et. al support and which is more stable and easier to install than tensorflow. However the missing pieces *that are DL-specific* should not live in SciPy. I understand that Keras wants to remain pure Python, for maintainability reasons. So why not write a small package that fills in the missing gaps?

**Comment by danFromTelAviv** [comment]
Thank you for your feedback. 
I assume you won't change your mind but I'll give it a try :)
So I started writing a small package like that to tackle just the really problematic functionalities ( in terms of runtime ) and I plan on using it / expending it for myself. 
This is a decent solution for me, but I think its not as good a solution for the community:
1. Hard to find small packages.
2. Hard to maintain/test well.
= Unlikely that python users will have the benefits I want to provide. 

I respect your view and I don't think DL stuff has to be a part of scipy as there are other packages that specialize in this. However, keep in mind that the key value of DL packages is training rather than inference. I think it would be reasonable for scipy to generalize for some DL applications since DL has become so significant in the scientific world. Again - not a must - but very nice to have. 

Maybe batches are a bit too specific - although batch processing is not just DL. But maybe there is some middle ground here for some functionalities that are intended for DL but don't sacrifice the generality of scipy as much. Maybe we can have a new conv function that is more geared towards computer vision uses like [sobel filters](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.sobel.html) that are obviously geared towards image processing.    

Something to think about...

**Comment by rgommers** [comment]
> ... This is a decent solution for me, but I think its not as good a solution for the community: ...

This is true in a way, but not really an argument for including it in SciPy (because that has costs for us and our non-DL users). 

>  I think it would be reasonable for scipy to generalize for some DL applications since DL has become so significant in the scientific world.

**Comment by rgommers** [comment]
Another thought: this issue is basically triggered by `keras` not wanting to add Cython code, right? A separate package could be easier to discover if it was an optional keras add-on, e.g. `pip install keras[numpy-backend]`

**Comment by danFromTelAviv** [comment]
Alright. Thank you for your thorough answer. 
That's a good idea with the naming stategy.
